### DISCUSSION TITLE
How can I do the equivalent of onload-script for my security wrapper project?

### DISCUSSION THREAD

**Discussion by ashward** [npm/feedback#293] [CLI] [answered]
I started a security wrapper project last year which allow-lists install scripts for security. It relies on the `onload-script` setting in a project `.npmrc` to hook into the npm process to intercept install scripts being invoked (by patching `child-process`).

I've just picked it up again (have been busy on work projects for several months!) and have obviously now realised that npm 7 has removed the `onload-script` function and as far as I can tell has no hooks into it, or anything to support what I need it to do (I've been through the docs and source and can't see anything I can hook into).

Am I missing something here? Is there a way to do what I need to do?

For yarn support I'm using `yarn-path` which works slightly differently, but does what I need it to. Maybe something more akin to that might be useful?

Here's the project I'm trying to update to support npm 7: https://github.com/ashward/jsinstallguard

Thanks!

Ash

**Comment by arcanis** [comment]
> For yarn support I'm using `yarn-path` which works slightly differently, but does what I need it to. Maybe something more akin to that might be useful?

Just as an fyi, Yarn supports build scripts allow-list / deny-list [natively](https://yarnpkg.com/configuration/manifest#dependenciesMeta.built).

**Reply by ashward** [reply]
Nice! That looks like it would cover most of the use cases I wanted to cover. I'll definitely have to have a closer look at that. 

Thanks for the heads-up!

**Comment by MylesBorins** [comment]
Pinging @npm/cli on this

**Reply by ashward** [reply]
Any news on this at all?

To be honest, `onload-script` wasn't ideal because there was no way of finding out the working directory in a cross-platform way (`${PWD}` works for Mac/Linux, but not on Windows), so I'm open to suggestions if there's another way to do what I need!

**Comment by darcyclarke** [comment, marked as answer]
@ashward I've [previously commented on this change & why we made it](https://github.com/npm/feedback/discussions/71#discussioncomment-116737); That said, we've actually been discussing expanding the _hooks_ you might be able to use for [tree mutations](https://github.com/npm/rfcs/issues/325) over on our [RFCs](https://github.com/npm/rfcs) repo - this may be something we introduce at some point but isn't immediately on our radar. I think your best bet is to try using either [**Hook Scripts**](https://docs.npmjs.com/cli/v7/using-npm/scripts#hook-scripts) or consume/interface with [`arborist`](https://github.com/npm/arborist)

**Reply by ashward** [reply]
Thanks Darcy.

I've had a look at those and unfortunately they don't do what I need. The Hook Scripts run in a separate child process so don't give an opportunity to patch the `child_process` module in the parent `npm` process, and also because they live in the `node_modules` directory they can't (or shouldn't) be checked in to source control, and one of my main aims is that the wrapper should be invoked without the user having to do anything else other than check the project out of source control. `.npmrc` and `.yarnrc` gave a mechanism to do this (although ideally something in `package.json` would be even better!)

Unfortunately I guess I'll have to not support npm v7 until I can figure something out 😞

**Reply by ljharb** [reply]
They can certainly still be checked into source control; you can add `!node_modules/.hooks` to `.gitignore`.

**Reply by ashward** [reply]
True. Although someone will soon enough do an `rm -rf node_modules` and unwittingly delete the hook!

**Reply by ashward** [reply]
I think I'll persevere with the hooks though and see if I can make it work reasonably, or come up with some other workaround.
